ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/MnecViolationCostEvaluator.java

```diff
@@ -0,0 +1,92 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.farao_community.farao.commons.Unit.AMPERE;
+import static com.farao_community.farao.commons.Unit.MEGAWATT;
+
+/**
+ * An evaluator that computes the virtual cost resulting from the violation of
+ * the MNEC minimum margin soft constraint
+ *
+ * @author Peter Mitri {@literal <peter.mitri at rte-france.com>}
+ */
+public class MnecViolationCostEvaluator implements CostEvaluator {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MnecViolationCostEvaluator.class);
+
+    private Unit unit;
+    private double mnecAcceptableMarginDiminution;
+    private double mnecViolationCost;
+
+    public MnecViolationCostEvaluator(Unit unit, double mnecAcceptableMarginDiminution, double mnecViolationCost) {
+        if ((unit != MEGAWATT) && (unit != AMPERE)) {
+            throw new NotImplementedException("MNEC violation cost is only implemented in MW and AMPERE units");
+        }
+        this.unit = unit;
+        this.mnecAcceptableMarginDiminution = mnecAcceptableMarginDiminution;
+        this.mnecViolationCost = mnecViolationCost;
+    }
+
+    @Override
+    public double getCost(RaoData raoData) {
+        if (Math.abs(mnecViolationCost) < 1e-10) {
+            return 0;
+        }
+        double totalMnecMarginViolation = 0;
+        boolean mnecsSkipped = false;
+        String initialVariantId =  raoData.getCrac().getExtension(ResultVariantManager.class).getPreOptimVariantId();
+        for (Cnec cnec : raoData.getCrac().getCnecs()) {
+            if (cnec.isMonitored()) {
+                double initialFlow = (unit == MEGAWATT) ? cnec.getExtension(CnecResultExtension.class).getVariant(initialVariantId).getFlowInMW()
+                        : cnec.getExtension(CnecResultExtension.class).getVariant(initialVariantId).getFlowInA();
+                if (Double.isNaN(initialFlow)) {
+                    // Sensitivity results are not available, skip cnec
+                    // (happens on search tree rao rootleaf evaluation)
+                    mnecsSkipped = true;
+                    continue;
+                }
+                double initialMargin = cnec.computeMargin(initialFlow, unit);
+                double newFlow = (unit == MEGAWATT) ? raoData.getSystematicSensitivityAnalysisResult().getReferenceFlow(cnec) :
+                        raoData.getSystematicSensitivityAnalysisResult().getReferenceIntensity(cnec);
+                double newMargin = cnec.computeMargin(newFlow, unit);
+                double convertedAcceptableMarginDiminution = mnecAcceptableMarginDiminution / getUnitConversionCoefficient(cnec, raoData);
+                totalMnecMarginViolation += Math.max(0, Math.min(0, initialMargin - convertedAcceptableMarginDiminution) - newMargin);
+            }
+        }
+        if (mnecsSkipped) {
+            LOGGER.warn("Some MNECs were skipped during violation cost evaluation because their sensitivity results were not available.");
```

Comment:
i would be more specific "initial flow results were not available"

Reply:
will do